// form-render/lib/widgets/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Card", {
  enumerable: true,
  get: function get() {
    return _boxcard.default;
  }
});
Object.defineProperty(exports, "CardList", {
  enumerable: true,
  get: function get() {
    return _listCard.default;
  }
});
Object.defineProperty(exports, "Checkbox", {
  enumerable: true,
  get: function get() {
    return _checkbox.default;
  }
});
Object.defineProperty(exports, "Checkboxes", {
  enumerable: true,
  get: function get() {
    return _checkboxes.default;
  }
});
Object.defineProperty(exports, "Collapse", {
  enumerable: true,
  get: function get() {
    return _boxCollapse.default;
  }
});
Object.defineProperty(exports, "Color", {
  enumerable: true,
  get: function get() {
    return _color.default;
  }
});
Object.defineProperty(exports, "DatePicker", {
  enumerable: true,
  get: function get() {
    return _date.default;
  }
});
Object.defineProperty(exports, "DateRange", {
  enumerable: true,
  get: function get() {
    return _dateRange.default;
  }
});
Object.defineProperty(exports, "DrawerList", {
  enumerable: true,
  get: function get() {
    return _listDrawer.default;
  }
});
Object.defineProperty(exports, "ErrorSchema", {
  enumerable: true,
  get: function get() {
    return _ErrorSchema.default;
  }
});
Object.defineProperty(exports, "Html", {
  enumerable: true,
  get: function get() {
    return _html.default;
  }
});
Object.defineProperty(exports, "ImageInput", {
  enumerable: true,
  get: function get() {
    return _imageInput.default;
  }
});
Object.defineProperty(exports, "Input", {
  enumerable: true,
  get: function get() {
    return _input.default;
  }
});
Object.defineProperty(exports, "InputNumber", {
  enumerable: true,
  get: function get() {
    return _number.default;
  }
});
Object.defineProperty(exports, "LineTitle", {
  enumerable: true,
  get: function get() {
    return _boxLineTitle.default;
  }
});
Object.defineProperty(exports, "MultiSelect", {
  enumerable: true,
  get: function get() {
    return _select.default;
  }
});
Object.defineProperty(exports, "PercentSlider", {
  enumerable: true,
  get: function get() {
    return _percentSlider.default;
  }
});
Object.defineProperty(exports, "Radio", {
  enumerable: true,
  get: function get() {
    return _radio.default;
  }
});
Object.defineProperty(exports, "Rate", {
  enumerable: true,
  get: function get() {
    return _rate.default;
  }
});
Object.defineProperty(exports, "Select", {
  enumerable: true,
  get: function get() {
    return _select.default;
  }
});
Object.defineProperty(exports, "SimpleList", {
  enumerable: true,
  get: function get() {
    return _listSimple.default;
  }
});
Object.defineProperty(exports, "Slider", {
  enumerable: true,
  get: function get() {
    return _slider.default;
  }
});
Object.defineProperty(exports, "SubInline", {
  enumerable: true,
  get: function get() {
    return _boxSubInline.default;
  }
});
Object.defineProperty(exports, "Switch", {
  enumerable: true,
  get: function get() {
    return _switch.default;
  }
});
Object.defineProperty(exports, "TabList", {
  enumerable: true,
  get: function get() {
    return _listTab.default;
  }
});
Object.defineProperty(exports, "TableList", {
  enumerable: true,
  get: function get() {
    return _listTable.default;
  }
});
Object.defineProperty(exports, "TextArea", {
  enumerable: true,
  get: function get() {
    return _textArea.default;
  }
});
Object.defineProperty(exports, "TimePicker", {
  enumerable: true,
  get: function get() {
    return _time.default;
  }
});
Object.defineProperty(exports, "TimeRange", {
  enumerable: true,
  get: function get() {
    return _timeRange.default;
  }
});
Object.defineProperty(exports, "TreeSelect", {
  enumerable: true,
  get: function get() {
    return _treeSelect.default;
  }
});
Object.defineProperty(exports, "Upload", {
  enumerable: true,
  get: function get() {
    return _upload.default;
  }
});
Object.defineProperty(exports, "UrlInput", {
  enumerable: true,
  get: function get() {
    return _urlInput.default;
  }
});
Object.defineProperty(exports, "VirtualList", {
  enumerable: true,
  get: function get() {
    return _listVirtual.default;
  }
});
Object.defineProperty(exports, "VoidTitle", {
  enumerable: true,
  get: function get() {
    return _voidTitle.default;
  }
});
var _input = _interopRequireDefault(require("./fields/input"));
var _number = _interopRequireDefault(require("./fields/number"));
var _textArea = _interopRequireDefault(require("./fields/textArea"));
var _select = _interopRequireDefault(require("./fields/select"));
var _switch = _interopRequireDefault(require("./fields/switch"));
var _radio = _interopRequireDefault(require("./fields/radio"));
var _rate = _interopRequireDefault(require("./fields/rate"));
var _treeSelect = _interopRequireDefault(require("./fields/treeSelect"));
var _checkbox = _interopRequireDefault(require("./fields/checkbox"));
var _checkboxes = _interopRequireDefault(require("./fields/checkboxes"));
var _color = _interopRequireDefault(require("./fields/color"));
var _date = _interopRequireDefault(require("./fields/date"));
var _dateRange = _interopRequireDefault(require("./fields/dateRange"));
var _time = _interopRequireDefault(require("./fields/time"));
var _timeRange = _interopRequireDefault(require("./fields/timeRange"));
var _imageInput = _interopRequireDefault(require("./fields/imageInput"));
var _urlInput = _interopRequireDefault(require("./fields/urlInput"));
var _slider = _interopRequireDefault(require("./fields/slider"));
var _upload = _interopRequireDefault(require("./fields/upload"));
var _html = _interopRequireDefault(require("./fields/html"));
var _percentSlider = _interopRequireDefault(require("./fields/percentSlider"));
var _boxcard = _interopRequireDefault(require("./boxcard"));
var _boxCollapse = _interopRequireDefault(require("./boxCollapse"));
var _boxSubInline = _interopRequireDefault(require("./boxSubInline"));
var _boxLineTitle = _interopRequireDefault(require("./boxLineTitle"));
var _listSimple = _interopRequireDefault(require("./listSimple"));
var _listCard = _interopRequireDefault(require("./listCard"));
var _listTable = _interopRequireDefault(require("./listTable"));
var _listDrawer = _interopRequireDefault(require("./listDrawer"));
var _listVirtual = _interopRequireDefault(require("./listVirtual"));
var _listTab = _interopRequireDefault(require("./listTab"));
var _voidTitle = _interopRequireDefault(require("./voidTitle"));
var _ErrorSchema = _interopRequireDefault(require("./ErrorSchema"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
