// @pdfme/schemas/dist/node/src/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDynamicHeightsForTable = exports.checkbox = exports.radioGroup = exports.select = exports.time = exports.date = exports.dateTime = exports.ellipse = exports.rectangle = exports.line = exports.barcodes = exports.table = exports.svg = exports.image = exports.multiVariableText = exports.text = exports.builtInPlugins = void 0;
const index_js_1 = __importDefault(require("./multiVariableText/index.js"));
exports.multiVariableText = index_js_1.default;
const index_js_2 = __importDefault(require("./text/index.js"));
exports.text = index_js_2.default;
const image_js_1 = __importDefault(require("./graphics/image.js"));
exports.image = image_js_1.default;
const svg_js_1 = __importDefault(require("./graphics/svg.js"));
exports.svg = svg_js_1.default;
const index_js_3 = __importDefault(require("./barcodes/index.js"));
exports.barcodes = index_js_3.default;
const line_js_1 = __importDefault(require("./shapes/line.js"));
exports.line = line_js_1.default;
const index_js_4 = __importDefault(require("./tables/index.js"));
exports.table = index_js_4.default;
const rectAndEllipse_js_1 = require("./shapes/rectAndEllipse.js");
Object.defineProperty(exports, "rectangle", { enumerable: true, get: function () { return rectAndEllipse_js_1.rectangle; } });
Object.defineProperty(exports, "ellipse", { enumerable: true, get: function () { return rectAndEllipse_js_1.ellipse; } });
const dateTime_js_1 = __importDefault(require("./date/dateTime.js"));
exports.dateTime = dateTime_js_1.default;
const date_js_1 = __importDefault(require("./date/date.js"));
exports.date = date_js_1.default;
const time_js_1 = __importDefault(require("./date/time.js"));
exports.time = time_js_1.default;
const index_js_5 = __importDefault(require("./select/index.js"));
exports.select = index_js_5.default;
const index_js_6 = __importDefault(require("./radioGroup/index.js"));
exports.radioGroup = index_js_6.default;
const index_js_7 = __importDefault(require("./checkbox/index.js"));
exports.checkbox = index_js_7.default;
const builtInPlugins = { Text: index_js_2.default };
exports.builtInPlugins = builtInPlugins;
// Export utility functions
var dynamicTemplate_js_1 = require("./tables/dynamicTemplate.js");
Object.defineProperty(exports, "getDynamicHeightsForTable", { enumerable: true, get: function () { return dynamicTemplate_js_1.getDynamicHeightsForTable; } });
